File: app/src/main/java/com/nvvi9/giphyme/db/GifPreviewDataStore.java
package com.nvvi9.giphyme.db;

import androidx.paging.PagingSource;

import com.nvvi9.giphyme.data.gifinfo.GifInfo;
import com.nvvi9.giphyme.data.preview.GifPreview;
import com.nvvi9.giphyme.db.dao.GifInfoDao;
import com.nvvi9.giphyme.db.dao.GifPreviewDao;

import java.util.List;
import java.util.stream.Collectors;

import javax.inject.Inject;

import io.reactivex.rxjava3.core.Completable;

public class GifPreviewDataStore {

    private final GifPreviewDao gifPreviewDao;
    private final GifInfoDao gifInfoDao;

    @Inject
    public GifPreviewDataStore(GifPreviewDao gifPreviewDao, GifInfoDao gifInfoDao) {
        this.gifPreviewDao = gifPreviewDao;
        this.gifInfoDao = gifInfoDao;
    }

    public Completable insert(List<GifPreview> gifPreviews) {
        List<String> previewIds = gifPreviews.stream().map(GifPreview::getId).collect(Collectors.toList());
        return gifInfoDao.getGifs(previewIds)
                .map(gifInfoList -> gifInfoList.stream().map(GifInfo::getId).collect(Collectors.toList()))
                .map(gifInfoIds -> gifPreviews.stream()
                        .map(gifPreview -> gifInfoIds.contains(gifPreview.getId())
                                ? new GifPreview(gifPreview.getId(), gifPreview.getTitle(), gifPreview.getPreviewUrl(), gifPreview.getPreviewUrl(), gifPreview.getHeight(), gifPreview.getWidth(), gifPreview.getSize(), true)
                                : gifPreview).collect(Collectors.toList()))
                .flatMapCompletable(gifPreviewDao::insert);
    }

    public Completable clear() {
        return gifPreviewDao.clear();
    }

    public PagingSource<Integer, GifPreview> getGifs() {
        return gifPreviewDao.getGifs();
    }

    public Completable markAsFavorite(String gifId, Boolean inFavorites) {
        return gifPreviewDao.updateSetInFavourites(gifId, inFavorites);
    }
}

File: app/src/main/java/com/nvvi9/giphyme/db/GifPreviewDataStore.kt
package com.nvvi9.giphyme.db

import androidx.paging.PagingSource
import com.nvvi9.giphyme.data.gifinfo.GifInfo
import com.nvvi9.giphyme.data.preview.GifPreview
import com.nvvi9.giphyme.db.dao.GifInfoDao
import com.nvvi9.giphyme.db.dao.GifPreviewDao
import io.reactivex.rxjava3.core.Completable
import java.util.stream.Collectors
import javax.inject.Inject

class GifPreviewDataStore @Inject constructor(
    private val gifPreviewDao: GifPreviewDao,
    private val gifInfoDao: GifInfoDao
) {
    fun insert(gifPreviews: List<GifPreview>): Completable =
        gifInfoDao.getGifs(gifPreviews.map { it.id })
            .map { gifInfoList -> gifInfoList.map { it.id } }
            .map { gifInfoIds ->
                gifPreviews
                    .map {
                        if (gifInfoIds.contains(it.id)) GifPreview(
                            it.id,
                            it.title,
                            it.previewUrl,
                            it.previewUrl,
                            it.height,
                            it.width,
                            it.size,
                            true
                        ) else it
                    }
            }
            .flatMapCompletable { gifPreviewDao.insert(it) }

    fun clear(): Completable = gifPreviewDao.clear()

    fun getGifs(): PagingSource<Int, GifPreview> =
        gifPreviewDao.getGifs()

    fun markAsFavorite(gifId: String, inFavorites: Boolean): Completable =
        gifPreviewDao.updateSetInFavourites(gifId, inFavorites)
}